OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs

```diff
@@ -67,20 +74,17 @@ class UpgradeState
 		}
 
 		readonly List<TimedUpgrade> timedUpgrades = new List<TimedUpgrade>();
-		readonly Lazy<Dictionary<string, UpgradeState>> upgrades;
 		readonly Dictionary<IUpgradable, int> levels = new Dictionary<IUpgradable, int>();
+		Dictionary<string, UpgradeState> upgrades;
 
-		public UpgradeManager(ActorInitializer init)
-		{
-			upgrades = Exts.Lazy(() =>
-			{
-				var ret = new Dictionary<string, UpgradeState>();
-				foreach (var up in init.Self.TraitsImplementing<IUpgradable>())
-					foreach (var t in up.UpgradeTypes)
-						ret.GetOrAdd(t).Traits.Add(up);
+		public UpgradeManager(ActorInitializer init) { }
```

Comment:
You could make `UpgradeManagerInfo : TraitInfo<UpgradeManager>, IRulesetLoaded` and then remove this empty ctor definition.